lib/mix/lib/mix/sync/pubsub.ex

```diff
@@ -0,0 +1,267 @@
+  # A publisher lists the directory, connects to each port and sends
+  # the message payload. When subscriber processes terminate, the
+  # files are left behind, so whenever a publisher fails to connect
+  # to one of the ports, it removes the corresponding file.
+  #
+  # We use a single socket for all the subscriptions. We create the
+  # socket lazily on the first subscription. Since the socket is
+  # shared, we include the hashed key in the broadcast payload, so
+  # that we can determine which subscribers to notify.
+
+  use GenServer
+
+  @opaque subscription :: %{socket: :gen_tcp.socket()}
+
+  @loopback {127, 0, 0, 1}
+  @listen_opts [:binary, ip: @loopback, packet: :raw, nodelay: true, backlog: 128, active: false]
+  @connect_opts [:binary, packet: :raw, nodelay: true, active: false]
+
+  @tag_data "mixpubsub"
+  @tag_data_size byte_size(@tag_data)
+
+  @name __MODULE__
+
+  @doc false
+  def start_link(_opts) do
+    GenServer.start_link(__MODULE__, {}, name: @name)
+  end
+
+  @doc """
+  Subscribes the caller to messages for `key`.
+
+  The messages are delivered to the caller's message queue.
+  """
+  @spec subscribe(String.t()) :: :ok
+  def subscribe(key) do
+    case GenServer.call(@name, {:subscribe, self(), key}, :infinity) do
+      :ok -> :ok
+      {:error, message} -> Mix.raise(message)
+    end
+  end
+
+  @doc """
+  Sends `message` to all processes subscribed to `key`.
+  """
+  @spec broadcast(String.t(), term()) :: :ok
+  def broadcast(key, message) do
+    hash = hash(key)
+    path = path(hash)
+
+    case File.ls(path) do
+      {:ok, names} ->
+        binary = :erlang.term_to_binary(message)
+
+        for "port_" <> port = name <- names do
+          port = String.to_integer(port)
+          port_path = Path.join(path, name)
+          _ = send_message(port_path, port, hash, binary)
+        end
+
+        :ok
+
+      {:error, :enoent} ->
+        :ok
+
+      {:error, reason} ->
+        raise File.Error, reason: reason, action: "list directory", path: path
+    end
+  end
+
+  defp send_message(port_path, port, hash, binary) do
+    # On Windows connecting to an unbound port takes a few seconds to
+    # fail, so instead we shortcut the check by attempting a listen,
+    # which succeeds or fails immediately. Note that `reuseaddr` here
+    # ensures that if the listening socket closed recently, we can
+    # immediately reclaim the same port.
+    #
+    # Also, if we manage to bind, it means the port is free and the
+    # subscription is no longer active, so we remove the file. It is
+    # important we remove the file while holding onto the port, so
+    # it is not claimed by a new subscriber as we remove the file.
+    #
+    # Finally, if we fail to connect, we could attempt to bind again
+    # and remove the file, but we just leave it up to the next send
+    # attempt.
+
+    case :gen_tcp.listen(port, [reuseaddr: true] ++ @listen_opts) do
+      {:ok, socket} ->
+        File.rm!(port_path)
```

Comment:
Could this fail due to a race condition? I.e. another process is trying to read the `port_path`? Maybe `File.rm` is better?

Reply:
Contrarily to the lock, these files are empty, the publisher does `File.ls!` and only the file names are relevant, so I  think we are good?

Reply:
In my experience Windows is very finicky with removing any files whatsoever. If there is any other OS process barely looking at it, removing the file will fail.